### ISSUE TITLE
Combine CVODES & IDAS into a single SUNDIALS lib

### ISSUE THREAD

**Issue by yizhang-cae** [stan-dev/math#779]
#### Summary:
use sundials instead of separate CVODES & IDAS to reduce library size. 

#### Description:
covdes and idas share nvector, sunlinsol and sunmtraix. Using a single sundials lib that contains these parts as well as cvodes & idas would reduce code size and maintenance effort.  

Specifically, Sundials distributes a single package consisting the different solvers. We can use this package instead of two libs (cvodes + idas). This removes the duplication of nvector, sunmatrix, and sublinsol libraries, cutting on size and linking/compiling time. See also discussion
https://github.com/stan-dev/math/pull/744#issuecomment-368215157

After combining the libraries, `lib` of `stan_math` will contain `sundials-3.1.0` with following structure
```
sundials-3.1.0/
├── include
│   ├── cvodes
│   ├── idas
│   ├── nvector
│   ├── sundials
│   ├── sunlinsol
│   └── sunmatrix
├── src
│   ├── cvodes
│   ├── idas
│   ├── nvec_ser
│   ├── sundials
│   ├── sunlinsol_band
│   ├── sunlinsol_dense
│   ├── sunlinsol_klu
│   ├── sunlinsol_lapackband
│   ├── sunlinsol_lapackdense
│   ├── sunlinsol_pcg
│   ├── sunlinsol_spbcgs
│   ├── sunlinsol_spfgmr
│   ├── sunlinsol_spgmr
│   ├── sunlinsol_sptfqmr
│   ├── sunlinsol_superlumt
│   ├── sunmat_band
│   ├── sunmat_dense
    └── sunmat_sparse

```
Changes to be made involve updating makefile & make/.

#### Current Version:
v2.17.1

**Comment by syclik** [comment]
Can you summarize the discussion?

And what's the proposed way of using the shared code?

**Comment by syclik** [comment]
This layout looks fine for Math and how it's going to interact with CmdStan.

@bgoodri, this look ok for RStan? Or are there any additional gotchas?
@ariddell and @ahartikainen, same questions for PyStan.

**Comment by ariddell** [comment]
I don't anticipate any problems with PyStan.
